File: iosApp/iosApp/Presentation/Settings/Account/AccountSettingsViewModel.swift
//
//  AccountSettingsViewModel.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/14.
//

import Foundation
import Combine
import Shared
import KMPNativeCoroutinesAsync

@MainActor
final class AccountSettingsViewModel: ObservableObject {
    private let getMyInfoUseCase: GetMyInfoUseCase

    private let observeCurrentUserUseCase: ObserveCurrentUserUseCase

    private let deleteAccountUseCase: DeleteAccountUseCase

    private let updateUserProfileUseCase: UpdateUserProfileUseCase

    @Published private(set) var uiState = AccountSettingsUiState.empty

    let event = PassthroughSubject<AccountSettingsEvent, Never>()

    private var tasks: [TaskKey: Task<Void, Never>] = [:]

    private func observeSession() {
        tasks[.observeSession]?.cancel()
        tasks[.observeSession] = Task {
            do {
                for try await _ in asyncSequence(for: observeCurrentUserUseCase.invoke()) {
                    self.loadAccountSettings()
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func loadAccountSettings() {
        uiState.isLoading = true
        uiState.errorMessage = nil

        tasks[.loadAccountSettings]?.cancel()
        tasks[.loadAccountSettings] = Task {
            do {
                let result = try await getMyInfoUseCase.invoke()
                if let success = result as? AppResultSuccess<AnyObject>,
                   let feed = success.data as? Shared.MyInfoFeed {
                    uiState.isLoading = false
                    uiState.errorMessage = nil
                    uiState.myInfoFeed = feed
                    uiState.nicknameInput = feed.user?.nickname ?? ""
                } else {
                    uiState.isLoading = false
                    uiState.errorMessage = String(localized: String.LocalizationValue("account_settings_error_load_failed"), table: "Localizable")
                }
            } catch {
                if Task.isCancelled { return }
                uiState.isLoading = false
                uiState.errorMessage = String(localized: String.LocalizationValue("account_settings_error_load_failed"), table: "Localizable")
            }
        }
    }

    private func emitMessage(_ message: String) {
        event.send(.showMessage(message))
    }

    private func saveUserInfo() {
        let nicknameInput = uiState.nicknameInput.trimmingCharacters(in: .whitespacesAndNewlines)
        let profileImage = uiState.myInfoFeed?.user?.profileImage

        if nicknameInput.isEmpty {
            emitMessage(String(localized: String.LocalizationValue("account_settings_message_enter_nickname"), table: "Localizable"))
        } else {
            uiState.isLoading = true
            uiState.errorMessage = nil

            tasks[.saveUserInfo]?.cancel()
            tasks[.saveUserInfo] = Task {
                do {
                    let result = try await updateUserProfileUseCase.invoke(nickname: nicknameInput, profileImage: profileImage)

                    if result is AppResultSuccess<AnyObject> {
                        uiState.isLoading = false
                        emitMessage(String(localized: String.LocalizationValue("account_settings_message_saved"), table: "Localizable"))
                        loadAccountSettings()
                    } else if let failure = result as? AppResultFailure {
                        let message = mapProfileUpdateFailureMessage(failure)
                        uiState.isLoading = false
                        uiState.errorMessage = nil
                        emitMessage(message)
                    } else {
                        uiState.isLoading = false
                        uiState.errorMessage = nil
                        emitMessage(String(localized: String.LocalizationValue("account_settings_error_profile_save_failed"), table: "Localizable"))
                    }
                } catch {
                    if Task.isCancelled { return }
                    uiState.isLoading = false
                    uiState.errorMessage = nil
                    emitMessage(String(localized: String.LocalizationValue("account_settings_error_profile_save_failed"), table: "Localizable"))
                }
            }
        }
    }

    private func openCastEdit() {
        guard let cast = uiState.myInfoFeed?.castDetail?.cast, !cast.id.isEmpty else {
            emitMessage(String(localized: String.LocalizationValue("account_settings_message_cast_profile_missing"), table: "Localizable"))
            return
        }
        event.send(.navigateToCastEdit(cafeId: cast.cafeId, castId: cast.id))
    }

    private func openChangePassword() {
        event.send(.navigateToChangePassword)
    }

    private func showDeleteDialog() {
        uiState.isDeleteDialogVisible = true
        uiState.deletePassword = ""
        uiState.deletePasswordErrorMessage = nil
    }

    private func dismissDeleteDialog() {
        uiState.isDeleteDialogVisible = false
        uiState.deletePassword = ""
        uiState.deletePasswordErrorMessage = nil
    }

    private func deleteAccount() {
        uiState.isLoading = true
        uiState.errorMessage = nil

        tasks[.deleteAccount]?.cancel()
        tasks[.deleteAccount] = Task {
            do {
                let result: AnyObject
                switch uiState.authProvider {
                case .email, .unknown:
                    let password = uiState.deletePassword.trimmingCharacters(in: .whitespacesAndNewlines)
                    if password.isEmpty {
                        uiState.isLoading = false
                        uiState.deletePasswordErrorMessage = String(localized: String.LocalizationValue("account_settings_delete_password_required"), table: "Localizable")
                        return
                    }
                    result = try await deleteAccountUseCase.invoke(
                        request: DeleteAccountRequest(
                            provider: .email,
                            password: password,
                            idToken: nil
                        )
                    )
                case .google, .kakao:
                    result = try await deleteAccountUseCase.invoke(
                        request: DeleteAccountRequest(
                            provider: uiState.authProvider,
                            password: nil,
                            idToken: nil
                        )
                    )
                case .apple:
                    uiState.isLoading = false
                    uiState.deletePasswordErrorMessage = String(localized: String.LocalizationValue("account_settings_error_apple_reauth_required"), table: "Localizable")
                    return
                default:
                    let password = uiState.deletePassword.trimmingCharacters(in: .whitespacesAndNewlines)
                    result = try await deleteAccountUseCase.invoke(
                        request: DeleteAccountRequest(
                            provider: .email,
                            password: password,
                            idToken: nil
                        )
                    )
                }
                if let failure = result as? AppResultFailure {
                    let message = mapDeleteFailureMessage(failure)
                    uiState.isLoading = false
                    uiState.errorMessage = nil
                    uiState.deletePasswordErrorMessage = message
                } else {
                    uiState.isLoading = false
                    uiState.errorMessage = nil
                    uiState.isDeleteRequested = true
                    uiState.isDeleteDialogVisible = false
                    uiState.deletePassword = ""
                    uiState.deletePasswordErrorMessage = nil
                    event.send(.navigateToMain)
                }
            } catch {
                if Task.isCancelled { return }
                uiState.isLoading = false
                uiState.errorMessage = String(localized: String.LocalizationValue("account_settings_error_delete_failed"), table: "Localizable")
                emitMessage(String(localized: String.LocalizationValue("account_settings_error_delete_failed"), table: "Localizable"))
            }
        }
    }

    private func deleteAccountWithAppleIdToken(_ idToken: String) {
        uiState.isLoading = true
        uiState.errorMessage = nil
        uiState.deletePasswordErrorMessage = nil

        tasks[.deleteAccount]?.cancel()
        tasks[.deleteAccount] = Task {
            do {
                let result = try await deleteAccountUseCase.invoke(
                    request: DeleteAccountRequest(
                        provider: .apple,
                        password: nil,
                        idToken: idToken
                    )
                )

                if let failure = result as? AppResultFailure {
                    let message = mapDeleteFailureMessage(failure)
                    uiState.isLoading = false
                    uiState.errorMessage = nil
                    uiState.deletePasswordErrorMessage = message
                } else {
                    uiState.isLoading = false
                    uiState.errorMessage = nil
                    uiState.isDeleteRequested = true
                    uiState.isDeleteDialogVisible = false
                    uiState.deletePassword = ""
                    uiState.deletePasswordErrorMessage = nil
                    event.send(.navigateToMain)
                }
            } catch {
                if Task.isCancelled { return }
                uiState.isLoading = false
                uiState.errorMessage = String(localized: String.LocalizationValue("account_settings_error_delete_failed"), table: "Localizable")
                emitMessage(String(localized: String.LocalizationValue("account_settings_error_delete_failed"), table: "Localizable"))
            }
        }
    }

    private func mapDeleteFailureMessage(_ failure: AppResultFailure) -> String {
        let rawError = String(describing: failure.error).uppercased()

        if rawError.contains("INVALID PASSWORD")
            || rawError.contains("INVALID_LOGIN_CREDENTIALS")
            || rawError.contains("INVALID_PASSWORD")
            || rawError.contains("EMAIL_NOT_FOUND") {
            return String(localized: String.LocalizationValue("account_settings_error_delete_invalid_password"), table: "Localizable")
        } else if rawError.contains("SOCIAL IDTOKEN IS REQUIRED") {
            return String(localized: String.LocalizationValue("account_settings_error_apple_reauth_required"), table: "Localizable")
        } else {
            return String(localized: String.LocalizationValue("account_settings_error_delete_failed"), table: "Localizable")
        }
    }

    private func mapProfileUpdateFailureMessage(_ failure: AppResultFailure) -> String {
        let rawError = String(describing: failure.error).uppercased()

        if rawError.contains("UNAUTHORIZED") {
            return String(localized: String.LocalizationValue("account_settings_error_session_expired"), table: "Localizable")
        } else if rawError.contains("VALIDATIONFAILED") {
            return String(localized: String.LocalizationValue("account_settings_message_enter_nickname"), table: "Localizable")
        } else {
            return String(localized: String.LocalizationValue("account_settings_error_profile_save_failed"), table: "Localizable")
        }
    }

    func onAction(_ action: AccountSettingsAction) {
        switch action {
        case .backTapped:
            event.send(.navigateBack)
        case .nicknameChanged(let value):
            uiState.nicknameInput = value
        case .saveUserInfoTapped:
            saveUserInfo()
        case .openCastEditTapped:
            openCastEdit()
        case .openChangePasswordTapped:
            openChangePassword()
        case .showDeleteDialogTapped:
            showDeleteDialog()
        case .dismissDeleteDialogTapped:
            dismissDeleteDialog()
        case .deletePasswordChanged(let value):
            uiState.deletePassword = value
            uiState.deletePasswordErrorMessage = nil
        case .deleteAccountTapped:
            deleteAccount()
        case .appleDeleteIdTokenReceived(let idToken):
            deleteAccountWithAppleIdToken(idToken)
        }
    }

    init(
        getMyInfoUseCase: GetMyInfoUseCase = KoinInitializerKt.resolveGetMyInfoUseCase(),
        observeCurrentUserUseCase: ObserveCurrentUserUseCase = KoinInitializerKt.resolveObserveCurrentUserUseCase(),
        deleteAccountUseCase: DeleteAccountUseCase = KoinInitializerKt.resolveDeleteAccountUseCase(),
        updateUserProfileUseCase: UpdateUserProfileUseCase = KoinInitializerKt.resolveUpdateUserProfileUseCase()
    ) {
        self.getMyInfoUseCase = getMyInfoUseCase
        self.observeCurrentUserUseCase = observeCurrentUserUseCase
        self.deleteAccountUseCase = deleteAccountUseCase
        self.updateUserProfileUseCase = updateUserProfileUseCase

        loadAccountSettings()
        observeSession()
    }

    deinit {
        tasks.values.forEach { task in
            task.cancel()
        }
        tasks.removeAll()
    }

    private enum TaskKey {
        case observeSession
        case loadAccountSettings
        case saveUserInfo
        case deleteAccount
    }
}

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/settings/account/AccountSettingsViewModel.kt
package com.hhp227.concafe.presentation.settings.account

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import concafe.composeapp.generated.resources.Res
import concafe.composeapp.generated.resources.account_settings_delete_password_required
import concafe.composeapp.generated.resources.account_settings_error_delete_apple_ios_only
import concafe.composeapp.generated.resources.account_settings_error_delete_failed
import concafe.composeapp.generated.resources.account_settings_error_delete_google_reauth_failed
import concafe.composeapp.generated.resources.account_settings_error_delete_invalid_password
import concafe.composeapp.generated.resources.account_settings_error_delete_kakao_reauth_failed
import concafe.composeapp.generated.resources.account_settings_error_load_failed
import concafe.composeapp.generated.resources.account_settings_error_profile_save_failed
import concafe.composeapp.generated.resources.account_settings_error_session_expired
import concafe.composeapp.generated.resources.account_settings_message_cast_profile_missing
import concafe.composeapp.generated.resources.account_settings_message_enter_nickname
import concafe.composeapp.generated.resources.account_settings_message_saved
import com.hhp227.concafe.domain.common.AppError
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.model.AuthProvider
import com.hhp227.concafe.domain.model.DeleteAccountRequest
import com.hhp227.concafe.domain.usecase.DeleteAccountUseCase
import com.hhp227.concafe.domain.usecase.GetMyInfoUseCase
import com.hhp227.concafe.domain.usecase.ObserveCurrentUserUseCase
import com.hhp227.concafe.domain.usecase.UpdateUserProfileUseCase
import com.hhp227.concafe.presentation.auth.signin.GoogleIdTokenProvider
import com.hhp227.concafe.presentation.auth.signin.KakaoIdTokenProvider
import org.jetbrains.compose.resources.StringResource
import org.jetbrains.compose.resources.getString

class AccountSettingsViewModel(
    private val getMyInfoUseCase: GetMyInfoUseCase,
    private val observeCurrentUserUseCase: ObserveCurrentUserUseCase,
    private val deleteAccountUseCase: DeleteAccountUseCase,
    private val updateUserProfileUseCase: UpdateUserProfileUseCase,
    private val googleIdTokenProvider: GoogleIdTokenProvider,
    private val kakaoIdTokenProvider: KakaoIdTokenProvider
) : ViewModel() {
    private val _uiState = MutableStateFlow(AccountSettingsUiState.empty())
    val uiState = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<AccountSettingsEvent>(replay = 0)
    val event = _event.asSharedFlow()

    private val jobs = mutableMapOf<TaskKey, Job>()

    private fun observeSession() {
        jobs[TaskKey.ObserveSession]?.cancel()
        jobs[TaskKey.ObserveSession] = viewModelScope.launch {
            observeCurrentUserUseCase.invoke().collectLatest {
                loadAccountSettings()
            }
        }
    }

    private fun loadAccountSettings() {
        _uiState.update { it.copy(isLoading = true, errorMessage = null) }

        jobs[TaskKey.LoadAccountSettings]?.cancel()
        jobs[TaskKey.LoadAccountSettings] = viewModelScope.launch {
            when (val result = getMyInfoUseCase.invoke()) {
                is AppResult.Success -> {
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        errorMessage = null,
                        myInfoFeed = result.data,
                        nicknameInput = result.data.user?.nickname.orEmpty()
                    )
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            errorMessage = getString(Res.string.account_settings_error_load_failed)
                        )
                    }
                }
            }
        }
    }

    private fun clickBack() {
        viewModelScope.launch {
            _event.emit(AccountSettingsEvent.NavigateBack)
        }
    }

    private fun clickSaveUserInfo() {
        val state = _uiState.value
        val nicknameInput = state.nicknameInput
        val profileImage = state.myInfoFeed?.user?.profileImage

        if (nicknameInput.isBlank()) {
            emitMessage(Res.string.account_settings_message_enter_nickname)
        } else {
            _uiState.update {
                it.copy(
                    isLoading = true,
                    errorMessage = null
                )
            }

            jobs[TaskKey.SaveUserInfo]?.cancel()
            jobs[TaskKey.SaveUserInfo] = viewModelScope.launch {
                when (val result = updateUserProfileUseCase.invoke(nicknameInput, profileImage)) {
                    is AppResult.Success -> {
                        _uiState.update {
                            it.copy(
                                isLoading = false,
                                errorMessage = null,
                                nicknameInput = result.data.nickname,
                                myInfoFeed = it.myInfoFeed?.copy(user = result.data)
                            )
                        }
                        emitMessage(Res.string.account_settings_message_saved)
                    }

                    is AppResult.Failure -> {
                        val message = mapProfileUpdateFailureMessage(result)

                        _uiState.update {
                            it.copy(
                                isLoading = false,
                                errorMessage = null
                            )
                        }
                        emitMessage(message)
                    }
                }
            }
        }
    }

    private fun clickOpenCastEdit() {
        val state = _uiState.value
        val cast = state.myInfoFeed?.castDetail?.cast
        if (cast?.id.isNullOrBlank()) {
            emitMessage(Res.string.account_settings_message_cast_profile_missing)
            return
        }
        viewModelScope.launch {
            _event.emit(
                AccountSettingsEvent.NavigateToCastEdit(
                    cafeId = cast!!.cafeId,
                    castId = cast.id
                )
            )
        }
    }

    private fun clickOpenChangePassword() {
        viewModelScope.launch {
            _event.emit(AccountSettingsEvent.NavigateToChangePassword)
        }
    }

    private fun clickShowDeleteDialog() {
        _uiState.update {
            it.copy(
                isDeleteDialogVisible = true,
                deletePassword = "",
                deletePasswordErrorMessage = null
            )
        }
    }

    private fun clickDismissDeleteDialog() {
        _uiState.update {
            it.copy(
                isDeleteDialogVisible = false,
                deletePassword = "",
                deletePasswordErrorMessage = null
            )
        }
    }

    private fun clickDeleteAccount() {
        val state = _uiState.value
        _uiState.update { it.copy(isLoading = true, errorMessage = null) }

        jobs[TaskKey.DeleteAccount]?.cancel()
        jobs[TaskKey.DeleteAccount] = viewModelScope.launch {
            val result = when (state.authProvider) {
                AuthProvider.EMAIL, AuthProvider.UNKNOWN -> {
                    if (state.deletePassword.isBlank()) {
                        _uiState.update {
                            it.copy(
                                isLoading = false,
                                deletePasswordErrorMessage = getString(Res.string.account_settings_delete_password_required)
                            )
                        }
                        return@launch
                    }
                    deleteAccountUseCase.invoke(
                        DeleteAccountRequest(
                            provider = AuthProvider.EMAIL,
                            password = state.deletePassword
                        )
                    )
                }
                AuthProvider.GOOGLE -> {
                    runCatching { googleIdTokenProvider.getGoogleIdToken() }
                        .fold(
                            onSuccess = { token ->
                                deleteAccountUseCase.invoke(
                                    DeleteAccountRequest(
                                        provider = AuthProvider.GOOGLE,
                                        idToken = token
                                    )
                                )
                            },
                            onFailure = {
                                AppResult.Failure(
                                    AppError.ValidationFailed("google reauth failed")
                                )
                            }
                        )
                }
                AuthProvider.KAKAO -> {
                    runCatching { kakaoIdTokenProvider.getKakaoAuthPayload().idToken }
                        .fold(
                            onSuccess = { token ->
                                deleteAccountUseCase.invoke(
                                    DeleteAccountRequest(
                                        provider = AuthProvider.KAKAO,
                                        idToken = token
                                    )
                                )
                            },
                            onFailure = {
                                AppResult.Failure(
                                    AppError.ValidationFailed("kakao reauth failed")
                                )
                            }
                        )
                }
                AuthProvider.APPLE -> {
                    AppResult.Failure(
                        AppError.ValidationFailed("apple delete not supported on compose")
                    )
                }
            }
            when (result) {
                is AppResult.Success -> {
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            errorMessage = null,
                            isDeleteRequested = true,
                            isDeleteDialogVisible = false,
                            deletePassword = ""
                        )
                    }
                    _event.emit(AccountSettingsEvent.NavigateToMain)
                }

                is AppResult.Failure -> {
                    val message = mapDeleteFailureMessage(result)

                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            errorMessage = null,
                            deletePasswordErrorMessage = message
                        )
                    }
                }
            }
        }
    }

    private fun emitMessage(message: String) {
        viewModelScope.launch {
            _event.emit(AccountSettingsEvent.ShowMessage(message))
        }
    }

    private fun emitMessage(message: StringResource) {
        viewModelScope.launch {
            _event.emit(AccountSettingsEvent.ShowMessage(getString(message)))
        }
    }

    private suspend fun mapDeleteFailureMessage(failure: AppResult.Failure): String {
        val rawError = failure.error.toString()
        val normalized = rawError.uppercase()

        return if (normalized.contains("INVALID PASSWORD")
            || normalized.contains("INVALID_LOGIN_CREDENTIALS")
            || normalized.contains("INVALID_PASSWORD")
            || normalized.contains("EMAIL_NOT_FOUND")
        ) {
            getString(Res.string.account_settings_error_delete_invalid_password)
        } else if (normalized.contains("GOOGLE REAUTH FAILED")) {
            getString(Res.string.account_settings_error_delete_google_reauth_failed)
        } else if (normalized.contains("KAKAO REAUTH FAILED")) {
            getString(Res.string.account_settings_error_delete_kakao_reauth_failed)
        } else if (normalized.contains("APPLE DELETE NOT SUPPORTED")) {
            getString(Res.string.account_settings_error_delete_apple_ios_only)
        } else {
            getString(Res.string.account_settings_error_delete_failed)
        }
    }

    private suspend fun mapProfileUpdateFailureMessage(failure: AppResult.Failure): String {
        val rawError = failure.error.toString().uppercase()

        return if (rawError.contains("UNAUTHORIZED")) {
            getString(Res.string.account_settings_error_session_expired)
        } else if (rawError.contains("VALIDATIONFAILED")) {
            getString(Res.string.account_settings_message_enter_nickname)
        } else {
            getString(Res.string.account_settings_error_profile_save_failed)
        }
    }

    fun onAction(action: AccountSettingsAction) {
        when (action) {
            AccountSettingsAction.ClickBack -> clickBack()
            is AccountSettingsAction.ChangeNickname -> _uiState.update { it.copy(nicknameInput = action.value) }
            AccountSettingsAction.ClickSaveUserInfo -> clickSaveUserInfo()
            AccountSettingsAction.ClickOpenCastEdit -> clickOpenCastEdit()
            AccountSettingsAction.ClickOpenChangePassword -> clickOpenChangePassword()
            AccountSettingsAction.ClickShowDeleteDialog -> clickShowDeleteDialog()
            AccountSettingsAction.ClickDismissDeleteDialog -> clickDismissDeleteDialog()
            is AccountSettingsAction.ChangeDeletePassword -> _uiState.update {
                it.copy(
                    deletePassword = action.value,
                    deletePasswordErrorMessage = null
                )
            }
            AccountSettingsAction.ClickDeleteAccount -> clickDeleteAccount()
        }
    }

    override fun onCleared() {
        jobs.values.forEach(Job::cancel)
        jobs.clear()
        super.onCleared()
    }

    init {
        loadAccountSettings()
        observeSession()
    }

    private enum class TaskKey {
        ObserveSession,
        LoadAccountSettings,
        SaveUserInfo,
        DeleteAccount
    }
}
